paasta_tools/kubernetes_tools.py

```diff
@@ -1376,31 +1378,71 @@ def list_deployments(
 
 
 @async_timeout()
-async def get_tail_lines_for_kubernetes_pod(
-    kube_client: KubeClient, pod: V1Pod, num_tail_lines: int
-) -> MutableMapping[str, List[str]]:
-    tail_lines_dict: MutableMapping[str, Any] = {
+async def get_tail_lines_for_kubernetes_container(
+    kube_client: KubeClient,
+    pod: V1Pod,
+    container: V1ContainerStatus,
+    num_tail_lines: int,
+) -> MutableMapping[str, Any]:
+    tail_lines: MutableMapping[str, Any] = {
         "stdout": [],
         "stderr": [],
         "error_message": "",
     }
-    for container in pod.spec.containers:
-        if container.name != HACHECK_POD_NAME:
-            try:
-                tail_lines_dict["stdout"].append(
-                    kube_client.core.read_namespaced_pod_log(
-                        name=pod.metadata.name,
-                        namespace="paasta",
-                        container=container.name,
-                        tail_lines=num_tail_lines,
+
+    if container.name != HACHECK_POD_NAME:
+        if container.state.waiting:
+            error = container.state.waiting.message
+        elif container.state.terminated:
+            error = container.state.terminated.message
+        error = error or ""  # if no errors, message will be None. convert to ""
+        tail_lines["error_message"] = error
+
+        try:
+            if num_tail_lines > 0:
+                log = kube_client.core.read_namespaced_pod_log(
+                    name=pod.metadata.name,
+                    namespace=pod.metadata.namespace,
+                    container=container.name,
+                    tail_lines=num_tail_lines,
+                )
+                tail_lines["stdout"].extend(log.split("\n"))
+        except ApiException as e:
+            # there is a potential race condition in which a pod's containers
+            # have not failed, but have when we get the container's logs. in this
+            # case, use the error from the exception, though it is less accurate.
+            if error == "":
+                body = json.loads(e.body)
+                error = body.get("message", "")
+            tail_lines["error_message"] = f"couldn't read stdout/stderr: '{error}'"
+
+    return tail_lines
+
+
+def format_tail_lines_for_kubernetes_pod(
+    pod_containers: Sequence, pod_name: str,
+) -> List[str]:
+    rows: List[str] = []
+    for container in pod_containers:
+        error_message = container.tail_lines.error_message
+        if error_message is not None:
```

Comment:
nit: if container.tail_lines.error_message is not None and replace error_message later? I don't think error_message make it much clear here.